dali/operators/decoder/audio/audio_decoder_op.cc

```diff
@@ -22,21 +23,85 @@ DALI_SCHEMA(AudioDecoder)
 This operator is a generic way of handling encoded data in DALI.
 It supports most of well-known audio formats (wav, flac, ogg).
 
-This operator produces two outputs:
-output[0]: batch of decoded data
+This operator produces two outputs: <br/>
+output[0]: batch of decoded data <br/>
 output[1]: batch of sampling rates [Hz]
 
 Sample rate (output[1]) at index `i` corresponds to sample (output[0]) at index `i`.
 On the event more metadata will appear, we reserve a right to change this behaviour.)code")
                 .NumInput(1)
                 .NumOutput(detail::kNumOutputs)
+                .AddOptionalArg(detail::kSampleRateName,
+                                "If specified, the output data will be resampled "
+                                "to the frequency provided [Hz]", -1)
                 .AddOptionalArg(
                         detail::kOutputTypeName,
                         "Type of the output data. Supports types: `INT16`, `INT32`, `FLOAT`",
                         DALI_INT16);
 
 DALI_REGISTER_OPERATOR(AudioDecoder, AudioDecoderCpu, CPU);
 
+std::vector<std::string> AudioDecoderCpu::files_names_ = {};
+std::vector<AudioMetadata> AudioDecoderCpu::sample_meta_ = {};
+std::vector<std::vector<float>> AudioDecoderCpu::intermediate_buffers_ = {};
+
+class AudioDecoderCpu::DecoderHelper {
+ public:
+  explicit DecoderHelper(AudioDecoderBase *decoder) : decoder_(decoder) {}
+
+
+  virtual void operator()(span<char> output_buffer, int &sample_rate, int sample_idx) = 0;
+
+ protected:
+  AudioDecoderBase *decoder_;
+};
+
+class AudioDecoderCpu::DirectDecoder : public AudioDecoderCpu::DecoderHelper {
+ public:
+  explicit DirectDecoder(AudioDecoderBase *decoder) : DecoderHelper(decoder) {}
+
+
+  void operator()(span<char> output_buffer, int &sample_rate, int sample_idx) override {
+    decoder_->Decode(output_buffer);
+    sample_rate = sample_meta_[sample_idx].sample_rate;
+  }
+
+
+ private:
+  using DecoderHelper::decoder_;
+};
+
+using T = float;
+
+class AudioDecoderCpu::ResamplingDecoder : public AudioDecoderCpu::DecoderHelper {
+ public:
+  ResamplingDecoder(AudioDecoderBase *decoder, int target_sample_rate) :
+          DecoderHelper(decoder), target_sample_rate_(target_sample_rate) {}
+
+
+  void operator()(span<char> output_buffer, int &sample_rate, int sample_idx) override {
+    intermediate_buffers_[sample_idx].resize(sample_meta_[sample_idx].length);
+    auto intermediate_buffer = make_span(intermediate_buffers_[sample_idx]);
+    auto intermediate_buffer_c = make_cspan(intermediate_buffers_[sample_idx]);
+    auto typed = (TypedAudioDecoderBase<T> *) (decoder_);
+    typed->DecodeTyped(intermediate_buffer);
+
+    // You don't need to downmix mono-channel stream
+    if (sample_meta_[sample_idx].channels > 1) {
+      Downmixing(intermediate_buffer, intermediate_buffer_c, sample_meta_[sample_idx].channels);
+    }
+
+    resampling::resample_sinc(reinterpret_cast<float *>(output_buffer.data()),
```

Comment:
If the output rate coincides with the input rate, don't resample, just copy! Resampling represents a computational load that should not be given away so lightly.